File: java/src/main/java/cn/wjybxx/dson/AbstractDsonWriter.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.io.DsonChunk;
import cn.wjybxx.dson.io.DsonIOException;
import cn.wjybxx.dson.text.INumberStyle;
import cn.wjybxx.dson.text.ObjectStyle;
import cn.wjybxx.dson.text.StringStyle;
import cn.wjybxx.dson.types.ObjectRef;
import cn.wjybxx.dson.types.OffsetTimestamp;
import com.google.protobuf.MessageLite;

import java.util.List;
import java.util.Objects;

/**
 * @author wjybxx
 * date - 2023/4/28
 */
public abstract class AbstractDsonWriter implements DsonWriter {

    protected final DsonWriterSettings settings;
    private Context context;
    private Context pooledContext; // 一个额外的缓存，用于写集合等减少上下文创建

    protected int recursionDepth;

    public AbstractDsonWriter(DsonWriterSettings settings) {
        this.settings = settings;
    }

    public DsonWriterSettings getSettings() {
        return settings;
    }

    protected Context getContext() {
        return context;
    }

    protected void setContext(Context context) {
        this.context = context;
    }

    protected Context getPooledContext() {
        return pooledContext;
    }

    protected void setPooledContext(Context pooledContext) {
        this.pooledContext = pooledContext;
    }

    @Override
    public void close() {
        context = null;
        pooledContext = null;
    }

    // region state

    @Override
    public DsonContextType getContextType() {
        return context.contextType;
    }

    @Override
    public boolean isAtName() {
        return context.state == DsonWriterState.NAME;
    }

    @Override
    public void writeName(String name) {
        Objects.requireNonNull(name, "name");
        Context context = this.context;
        if (context.state != DsonWriterState.NAME) {
            throw invalidState(List.of(DsonWriterState.NAME), context.state);
        }
        context.curName = name;
        context.state = DsonWriterState.VALUE;
        doWriteName(name);
    }

    /** 执行{@link #writeName(String)}时调用 */
    protected void doWriteName(String name) {

    }

    protected void advanceToValueState(String name) {
        Context context = this.context;
        if (context.state == DsonWriterState.NAME) {
            writeName(name);
        }
        if (context.state != DsonWriterState.VALUE) {
            throw invalidState(List.of(DsonWriterState.VALUE), context.state);
        }
    }

    protected void ensureValueState(Context context) {
        if (context.state != DsonWriterState.VALUE) {
            throw invalidState(List.of(DsonWriterState.VALUE), context.state);
        }
    }

    protected void setNextState() {
        switch (context.contextType) {
            case OBJECT, HEADER -> context.setState(DsonWriterState.NAME);
            case TOP_LEVEL, ARRAY -> context.setState(DsonWriterState.VALUE);
        }
    }

    private DsonIOException invalidState(List<DsonWriterState> expected, DsonWriterState state) {
        return DsonIOException.invalidState(context.contextType, expected, state);
    }
    // endregion

    // region 简单值
    @Override
    public void writeInt32(String name, int value, WireType wireType, INumberStyle style) {
        advanceToValueState(name);
        doWriteInt32(value, wireType, style);
        setNextState();
    }

    @Override
    public void writeInt64(String name, long value, WireType wireType, INumberStyle style) {
        advanceToValueState(name);
        doWriteInt64(value, wireType, style);
        setNextState();
    }

    @Override
    public void writeFloat(String name, float value, INumberStyle style) {
        advanceToValueState(name);
        doWriteFloat(value, style);
        setNextState();
    }

    @Override
    public void writeDouble(String name, double value, INumberStyle style) {
        advanceToValueState(name);
        doWriteDouble(value, style);
        setNextState();
    }

    @Override
    public void writeBoolean(String name, boolean value) {
        advanceToValueState(name);
        doWriteBool(value);
        setNextState();
    }

    @Override
    public void writeString(String name, String value, StringStyle style) {
        Objects.requireNonNull(value);
        advanceToValueState(name);
        doWriteString(value, style);
        setNextState();
    }

    @Override
    public void writeNull(String name) {
        advanceToValueState(name);
        doWriteNull();
        setNextState();
    }

    @Override
    public void writeBinary(String name, DsonBinary dsonBinary) {
        Objects.requireNonNull(dsonBinary);
        advanceToValueState(name);
        doWriteBinary(dsonBinary);
        setNextState();
    }

    @Override
    public void writeBinary(String name, int type, DsonChunk chunk) {
        Objects.requireNonNull(chunk);
        Dsons.checkSubType(type);
        Dsons.checkBinaryLength(chunk.getLength());
        advanceToValueState(name);
        doWriteBinary(type, chunk);
        setNextState();
    }

    @Override
    public void writeExtInt32(String name, DsonExtInt32 value, WireType wireType, INumberStyle style) {
        Objects.requireNonNull(value);
        advanceToValueState(name);
        doWriteExtInt32(value, wireType, style);
        setNextState();
    }

    @Override
    public void writeExtInt64(String name, DsonExtInt64 value, WireType wireType, INumberStyle style) {
        Objects.requireNonNull(value);
        advanceToValueState(name);
        doWriteExtInt64(value, wireType, style);
        setNextState();
    }

    @Override
    public void writeExtDouble(String name, DsonExtDouble value, INumberStyle style) {
        Objects.requireNonNull(value);
        advanceToValueState(name);
        doWriteExtDouble(value, style);
        setNextState();
    }

    @Override
    public void writeExtString(String name, DsonExtString value, StringStyle style) {
        Objects.requireNonNull(value);
        advanceToValueState(name);
        doWriteExtString(value, style);
        setNextState();
    }

    @Override
    public void writeRef(String name, ObjectRef objectRef) {
        Objects.requireNonNull(objectRef);
        advanceToValueState(name);
        doWriteRef(objectRef);
        setNextState();
    }

    @Override
    public void writeTimestamp(String name, OffsetTimestamp timestamp) {
        Objects.requireNonNull(timestamp);
        advanceToValueState(name);
        doWriteTimestamp(timestamp);
        setNextState();
    }

    protected abstract void doWriteInt32(int value, WireType wireType, INumberStyle style);

    protected abstract void doWriteInt64(long value, WireType wireType, INumberStyle style);

    protected abstract void doWriteFloat(float value, INumberStyle style);

    protected abstract void doWriteDouble(double value, INumberStyle style);

    protected abstract void doWriteBool(boolean value);

    protected abstract void doWriteString(String value, StringStyle style);

    protected abstract void doWriteNull();

    protected abstract void doWriteBinary(DsonBinary binary);

    protected abstract void doWriteBinary(int type, DsonChunk chunk);

    protected abstract void doWriteExtInt32(DsonExtInt32 extInt32, WireType wireType, INumberStyle style);

    protected abstract void doWriteExtInt64(DsonExtInt64 extInt64, WireType wireType, INumberStyle style);

    protected abstract void doWriteExtDouble(DsonExtDouble extDouble, INumberStyle style);

    protected abstract void doWriteExtString(DsonExtString extString, StringStyle style);

    protected abstract void doWriteRef(ObjectRef objectRef);

    protected abstract void doWriteTimestamp(OffsetTimestamp timestamp);

    // endregion

    // region 容器
    @Override
    public void writeStartArray(ObjectStyle style) {
        writeStartContainer(DsonContextType.ARRAY, DsonType.ARRAY, style);
    }

    @Override
    public void writeEndArray() {
        writeEndContainer(DsonContextType.ARRAY, DsonWriterState.VALUE);
    }

    @Override
    public void writeStartObject(ObjectStyle style) {
        writeStartContainer(DsonContextType.OBJECT, DsonType.OBJECT, style);
    }

    @Override
    public void writeEndObject() {
        writeEndContainer(DsonContextType.OBJECT, DsonWriterState.NAME);
    }

    @Override
    public void writeStartHeader(ObjectStyle style) {
        // object下默认是name状态
        Context context = this.context;
        if (context.contextType == DsonContextType.OBJECT && context.state == DsonWriterState.NAME) {
            context.setState(DsonWriterState.VALUE);
        }
        writeStartContainer(DsonContextType.HEADER, DsonType.HEADER, style);
    }

    @Override
    public void writeEndHeader() {
        writeEndContainer(DsonContextType.HEADER, DsonWriterState.NAME);
    }

    private void writeStartContainer(DsonContextType contextType, DsonType dsonType, ObjectStyle style) {
        Objects.requireNonNull(style);
        if (recursionDepth >= settings.recursionLimit) {
            throw DsonIOException.recursionLimitExceeded();
        }
        Context context = this.context;
        autoStartTopLevel(context);
        ensureValueState(context);
        doWriteStartContainer(contextType, dsonType, style);
        setNextState(); // 设置新上下文状态
    }

    private void writeEndContainer(DsonContextType contextType, DsonWriterState expectedState) {
        Context context = this.context;
        checkEndContext(context, contextType, expectedState);
        doWriteEndContainer();
        setNextState(); // parent前进一个状态
    }

    protected void autoStartTopLevel(Context context) {
        if (context.contextType == DsonContextType.TOP_LEVEL
                && context.state == DsonWriterState.INITIAL) {
            context.setState(DsonWriterState.VALUE);
        }
    }

    protected void checkEndContext(Context context, DsonContextType contextType, DsonWriterState state) {
        if (context.contextType != contextType) {
            throw DsonIOException.contextError(contextType, context.contextType);
        }
        if (context.state != state) {
            throw invalidState(List.of(state), context.state);
        }
    }

    /** 写入类型信息，创建新上下文，压入上下文 */
    protected abstract void doWriteStartContainer(DsonContextType contextType, DsonType dsonType, ObjectStyle style);

    /** 弹出上下文 */
    protected abstract void doWriteEndContainer();

    // endregion
    // region sp

    @Override
    public void writeMessage(String name, int binaryType, MessageLite messageLite) {
        Dsons.checkSubType(binaryType);
        advanceToValueState(name);
        doWriteMessage(binaryType, messageLite);
        setNextState();
    }

    @Override
    public void writeValueBytes(String name, DsonType type, byte[] data) {
        DsonReaderUtils.checkWriteValueAsBytes(type);
        advanceToValueState(name);
        doWriteValueBytes(type, data);
        setNextState();
    }

    @Override
    public Object attach(Object userData) {
        return context.attach(userData);
    }

    @Override
    public Object attachment() {
        return context.userData;
    }

    protected abstract void doWriteMessage(int binaryType, MessageLite messageLite);

    protected abstract void doWriteValueBytes(DsonType type, byte[] data);

    // endregion

    // region context

    protected static class Context {

        public Context parent;
        public DsonContextType contextType;
        public DsonType dsonType; // 用于在Object/Array模式下写入内置数据结构
        public DsonWriterState state = DsonWriterState.INITIAL;
        public String curName;
        public Object userData;

        public Context() {
        }

        public Context init(Context parent, DsonContextType contextType, DsonType dsonType) {
            this.parent = parent;
            this.contextType = contextType;
            this.dsonType = dsonType;
            return this;
        }

        public void reset() {
            parent = null;
            contextType = null;
            dsonType = null;
            state = DsonWriterState.INITIAL;
            curName = null;
            userData = null;
        }

        public Object attach(Object userData) {
            Object r = this.userData;
            this.userData = userData;
            return r;
        }

        /** 方便查看赋值的调用 */
        public void setState(DsonWriterState state) {
            this.state = state;
        }

        public Context getParent() {
            return parent;
        }

    }

}

File: csharp/Dson/src/AbstractDsonWriter.cs
#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using Wjybxx.Dson.IO;
using Wjybxx.Dson.Text;
using Wjybxx.Dson.Types;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

public abstract class AbstractDsonWriter<TName> : IDsonWriter<TName> where TName : IEquatable<TName>
{
    protected readonly DsonWriterSettings Settings;
    private const int PoolSize = 4;

#nullable disable
    internal Context _context;
    private Stack<Context> _contextPool = new(PoolSize);
    protected int _recursionDepth = 0; // 当前递归深度
#nullable enable

    protected AbstractDsonWriter(DsonWriterSettings settings) {
        this.Settings = settings ?? throw new ArgumentNullException(nameof(settings));
    }

    public DsonWriterSettings WriterSettings => Settings;

    protected virtual Context GetContext() {
        return _context;
    }

    protected void SetContext(Context context) {
        this._context = context;
    }

    protected abstract Context NewContext();

    protected Context RentContext() {
        if (_contextPool.TryPop(out Context? context)) {
            return context;
        }
        return NewContext();
    }

    protected void ReturnContext(Context context) {
        context.Reset();
        if (_contextPool.Count < PoolSize) {
            _contextPool.Push(context);
        }
    }

    public abstract void Flush();

    public virtual void Dispose() {
        _context = null!;
        _contextPool.Clear();
        _contextPool = null!;
    }

    #region state

    public DsonContextType ContextType => _context.contextType;

    public TName CurrentName {
        get {
            Context context = this._context;
            if (context.state != DsonWriterState.Value) {
                throw InvalidState(DsonInternals.NewList(DsonWriterState.Value), context.state);
            }
            return context.curName;
        }
    }

    public bool IsAtName => _context.state == DsonWriterState.Name;

    public void WriteName(TName name) {
        if (name == null) throw new ArgumentNullException(nameof(name));
        Context context = this._context;
        if (context.state != DsonWriterState.Name) {
            throw InvalidState(DsonInternals.NewList(DsonWriterState.Name), context.state);
        }
        context.curName = name;
        context.state = DsonWriterState.Value;
        DoWriteName(name);
    }

    /** 执行{@link #WriteName(String)}时调用 */
    protected void DoWriteName(TName name) {
    }

    protected void AdvanceToValueState(TName name) {
        Context context = this._context;
        if (context.state == DsonWriterState.Name) {
            WriteName(name);
        }
        if (context.state != DsonWriterState.Value) {
            throw InvalidState(DsonInternals.NewList(DsonWriterState.Value), context.state);
        }
    }

    protected void EnsureValueState(Context context) {
        if (context.state != DsonWriterState.Value) {
            throw InvalidState(DsonInternals.NewList(DsonWriterState.Value), context.state);
        }
    }

    protected void SetNextState() {
        switch (_context.contextType) {
            case DsonContextType.Object:
            case DsonContextType.Header: {
                _context.SetState(DsonWriterState.Name);
                break;
            }
            case DsonContextType.TopLevel:
            case DsonContextType.Array: {
                _context.SetState(DsonWriterState.Value);
                break;
            }
            default: throw new InvalidOperationException();
        }
    }

    private DsonIOException InvalidState(List<DsonWriterState> expected, DsonWriterState state) {
        return DsonIOException.InvalidState(_context.contextType, expected, state);
    }

    #endregion

    #region 简单值

    public void WriteInt32(TName name, int value, WireType wireType, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteInt32(value, wireType, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteInt64(TName name, long value, WireType wireType, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteInt64(value, wireType, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteFloat(TName name, float value, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteFloat(value, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteDouble(TName name, double value, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteDouble(value, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteBoolean(TName name, bool value) {
        AdvanceToValueState(name);
        DoWriteBool(value);
        SetNextState();
    }

    public void WriteString(TName name, string value, StringStyle style = StringStyle.Auto) {
        if (value == null) throw new ArgumentNullException(nameof(value));
        AdvanceToValueState(name);
        DoWriteString(value, style);
        SetNextState();
    }

    public void WriteNull(TName name) {
        AdvanceToValueState(name);
        DoWriteNull();
        SetNextState();
    }

    public void WriteBinary(TName name, Binary binary) {
        AdvanceToValueState(name);
        DoWriteBinary(binary);
        SetNextState();
    }

    public void WriteBinary(TName name, int type, DsonChunk chunk) {
        Dsons.CheckSubType(type);
        Dsons.CheckBinaryLength(chunk.Length);
        AdvanceToValueState(name);
        DoWriteBinary(type, chunk);
        SetNextState();
    }

    public void WriteExtInt32(TName name, ExtInt32 extInt32, WireType wireType, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteExtInt32(extInt32, wireType, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteExtInt64(TName name, ExtInt64 extInt64, WireType wireType, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteExtInt64(extInt64, wireType, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteExtDouble(TName name, ExtDouble extDouble, INumberStyle? style = null) {
        AdvanceToValueState(name);
        DoWriteExtDouble(extDouble, style ?? NumberStyles.Simple);
        SetNextState();
    }

    public void WriteExtString(TName name, ExtString extString, StringStyle style = StringStyle.Auto) {
        AdvanceToValueState(name);
        DoWriteExtString(extString, style);
        SetNextState();
    }

    public void WriteRef(TName name, ObjectRef objectRef) {
        AdvanceToValueState(name);
        DoWriteRef(objectRef);
        SetNextState();
    }

    public void WriteTimestamp(TName name, OffsetTimestamp timestamp) {
        AdvanceToValueState(name);
        DoWriteTimestamp(timestamp);
        SetNextState();
    }

    protected abstract void DoWriteInt32(int value, WireType wireType, INumberStyle style);

    protected abstract void DoWriteInt64(long value, WireType wireType, INumberStyle style);

    protected abstract void DoWriteFloat(float value, INumberStyle style);

    protected abstract void DoWriteDouble(double value, INumberStyle style);

    protected abstract void DoWriteBool(bool value);

    protected abstract void DoWriteString(string value, StringStyle style);

    protected abstract void DoWriteNull();

    protected abstract void DoWriteBinary(Binary binary);

    protected abstract void DoWriteBinary(int type, DsonChunk chunk);

    protected abstract void DoWriteExtInt32(ExtInt32 extInt32, WireType wireType, INumberStyle style);

    protected abstract void DoWriteExtInt64(ExtInt64 extInt64, WireType wireType, INumberStyle style);

    protected abstract void DoWriteExtDouble(ExtDouble extDouble, INumberStyle style);

    protected abstract void DoWriteExtString(ExtString extString, StringStyle style);

    protected abstract void DoWriteRef(ObjectRef objectRef);

    protected abstract void DoWriteTimestamp(OffsetTimestamp timestamp);

    #endregion

    #region 容器

    public void WriteStartArray(ObjectStyle style) {
        WriteStartContainer(DsonContextType.Array, DsonType.Array, style);
    }

    public void WriteEndArray() {
        WriteEndContainer(DsonContextType.Array, DsonWriterState.Value);
    }

    public void WriteStartObject(ObjectStyle style) {
        WriteStartContainer(DsonContextType.Object, DsonType.Object, style);
    }

    public void WriteEndObject() {
        WriteEndContainer(DsonContextType.Object, DsonWriterState.Name);
    }

    public void WriteStartHeader(ObjectStyle style) {
        // object下默认是name状态
        Context context = this._context;
        if (context.contextType == DsonContextType.Object && context.state == DsonWriterState.Name) {
            context.SetState(DsonWriterState.Value);
        }
        WriteStartContainer(DsonContextType.Header, DsonType.Header, style);
    }

    public void WriteEndHeader() {
        WriteEndContainer(DsonContextType.Header, DsonWriterState.Name);
    }

    private void WriteStartContainer(DsonContextType contextType, DsonType dsonType, ObjectStyle style) {
        if (_recursionDepth >= Settings.RecursionLimit) {
            throw DsonIOException.RecursionLimitExceeded();
        }
        Context context = this._context;
        AutoStartTopLevel(context);
        EnsureValueState(context);
        DoWriteStartContainer(contextType, dsonType, style);
        SetNextState(); // 设置新上下文状态
    }

    private void WriteEndContainer(DsonContextType contextType, DsonWriterState expectedState) {
        Context context = this._context;
        CheckEndContext(context, contextType, expectedState);
        DoWriteEndContainer();
        SetNextState(); // parent前进一个状态
    }

    protected void AutoStartTopLevel(Context context) {
        if (context.contextType == DsonContextType.TopLevel
            && context.state == DsonWriterState.Initial) {
            context.SetState(DsonWriterState.Value);
        }
    }

    protected void CheckEndContext(Context context, DsonContextType contextType, DsonWriterState state) {
        if (context.contextType != contextType) {
            throw DsonIOException.ContextError(contextType, context.contextType);
        }
        if (context.state != state) {
            throw InvalidState(DsonInternals.NewList(state), context.state);
        }
    }

    /** 写入类型信息，创建新上下文，压入上下文 */
    protected abstract void DoWriteStartContainer(DsonContextType contextType, DsonType dsonType, ObjectStyle style);

    /** 弹出上下文 */
    protected abstract void DoWriteEndContainer();

    #endregion

    #region 特殊

    public virtual void WriteSimpleHeader(string clsName) {
        if (clsName == null) throw new ArgumentNullException(nameof(clsName));
        IDsonWriter<string> textWrite = (IDsonWriter<string>)this;
        textWrite.WriteStartHeader();
        textWrite.WriteString(DsonHeaders.NamesClassName, clsName, StringStyle.AutoQuote);
        textWrite.WriteEndHeader();
    }

    public void WriteValueBytes(TName name, DsonType type, byte[] data) {
        DsonReaderUtils.CheckWriteValueAsBytes(type);
        AdvanceToValueState(name);
        DoWriteValueBytes(type, data);
        SetNextState();
    }

    public object? Attach(object userData) {
        return _context.Attach(userData);
    }

    public object Attachment() {
        return _context.userData;
    }

    protected abstract void DoWriteValueBytes(DsonType type, byte[] data);

    #endregion

    #region contxt

#nullable enable
    protected internal abstract class Context
    {
#nullable disable
        protected internal Context parent;
        protected internal DsonContextType contextType;
        protected internal DsonType dsonType; // 用于在Object/Array模式下写入内置数据结构
        protected internal DsonWriterState state = DsonWriterState.Initial;
        protected internal TName curName;
        protected internal object userData;
#nullable enable

        public Context() {
        }

        public Context Init(Context? parent, DsonContextType contextType, DsonType dsonType) {
            this.parent = parent;
            this.contextType = contextType;
            this.dsonType = dsonType;
            return this;
        }

        public virtual void Reset() {
            parent = null;
            contextType = default;
            dsonType = DsonTypes.Invalid;
            state = default;
            curName = default;
            userData = null;
        }

        public object? Attach(object? userData) {
            object? r = this.userData;
            this.userData = userData;
            return r;
        }

        /** 方便查看赋值的调用 */
        public void SetState(DsonWriterState state) {
            this.state = state;
        }

        /** 子类可隐藏 */
        public Context Parent => parent;
    }

    #endregion
}